### DISCUSSION TITLE
Which UI components would you like to see in transaction insights?

### DISCUSSION THREAD

**Discussion by Montoya** [MetaMask/snaps#1859] [General]
Hello Snap devs, we would like to get your feedback on which UI components you would like to be able to use in transaction insight views. On top of what is currently possible, we are adding support for the following: 

- Links
- SVG images (non interactive)

We are also planning to add: 

- Key / value pairs, like this example: 
   <img width="324" alt="Screenshot 2023-10-17 at 11 10 44 AM" src="https://github.com/MetaMask/snaps/assets/19909/ecd033e3-4365-4f15-8728-f4720e693e62">

What else would you like to see? Lists? Tables? Something more complex? Feel free to include examples.

**Comment by mindofmar** [comment]
Awesome additions. Images (PNGs specifically) for state changes would be a huge win. Also tooltips are always a nice to have.

![image](https://github.com/MetaMask/snaps/assets/72634565/1989761c-f17b-4992-8dde-16c45a927128)

**Reply by eriknson** [reply]
Why do you specifically need PNGs over e.g. SVGs?

**Reply by mindofmar** [reply]
Mainly because within a lot of web3 metadata, the image URLs are served as PNG

**Reply by eriknson** [reply]
Ok, that makes sense, I’ll discuss including PNG support with the team.

**Comment by azep-ninja** [comment]
Tooltips for sure as already mentioned, and disclore/accordion to show/hide content so if data is displayed that maybe long it can be expandable without immidiately taking up the full windows and scrolling forever.

Default
![image](https://github.com/MetaMask/snaps/assets/142059473/8fea158b-6dbb-4b4f-8a42-1a421f9fe9fc)

Expanded
![image](https://github.com/MetaMask/snaps/assets/142059473/bb86c7f6-7651-4519-b19b-dd52d2231851)

**Comment by gmkung** [comment]
Here are the things on the wishlist from team Kleros!
- Links (already mentioned)
- Tooltips (already mentioned)
- PNG images (also mentioned)
- Richer formatting (color in text, emoticons 💡 🚫 🆗 )
- Ability to display Snap in signature and token transfer flows (already in roadmap I believe)

**Reply by eriknson** [reply]
Thanks for this input, much appreciated.

Emoticons is already available, let us know if you have any issues using them though.

**Reply by eriknson** [reply]
Any specific reason to wanting PNGs specifically? 

We enjoy vectorized images, so our initial scope (as of now) will be SVGs. Would this work for your use case?

**Reply by gmkung** [reply]
Most of the images we curate through our [token curated registries](https://curate.kleros.io/tcr/100/0xeE1502e29795Ef6C2D60F8D7120596abE3baD990) are in PNG as that's the most commonly available forma. We can convert them to SVG but it would add complexity to the Snap.

**Reply by eriknson** [reply]
Makes sense, will discuss this w/ the team

**Comment by vidorge** [comment]
Would be cool if copyable elements could be collapsed by default with only first (or few) lines showing with the possibility to expand.
This would allow displaying more meaningful data in the already small window.

<img width="472" alt="image" src="https://github.com/MetaMask/snaps/assets/989725/cfe91cb6-3304-49f4-8c6c-85d9865cff2b">

**Comment by juanforta** [comment]
We agree with all the suggestions proposed. On top, would be great a setting that would allow the user to automatically add the 3rd party snap data within the "Details" window. Is integrating snaps with the transaction flow on the roadmap?

**Comment by dzimiks** [comment]
- Accordion
- Rich Markdown support
- Adding custom styling (CSS, tailwind, anything would be helpful)
- Ability to embed/render charts - pie, bar, line, etc.
- Writing custom React/HTML code
- Alert/Callout component
- Badge/Button component
- Dropdown Menu with actionable items/links
- Input/Select/Checkbox/Forms - it can be useful if someone needs to enter third-party API key or similar

**Comment by danfinlay** [comment]
- List item picker (let a user select from a list of provided items)
- Pet name renderer (let MetaMask insert its best name for an item that the snap has reference to, like an ethereum address being substituted for an address book name).

**Comment by wangkui0508** [comment]
I'd like to see a safe subset of HTML and CSS. Many UI components can be made using such a subset. Just like Qt does: https://doc.qt.io/qt-6/richtext-html-subset.html

**Comment by Alex-distributed-lab** [comment]
Button components
Different font sizes (at least basic S, M, L, XL)
Subtabs like on the main Metamask page
Cards with attachable elements (pictures, text)
Scrollbars (including horizontal)
Dropdowns
<img width="352" alt="Screenshot 2023-11-14 at 20 04 20" src="https://github.com/MetaMask/snaps/assets/68012282/4a8545bb-836e-4df4-aa1f-1bedcffc0580">
<img width="477" alt="Screenshot 2023-11-14 at 20 13 09" src="https://github.com/MetaMask/snaps/assets/68012282/32621495-a82c-4ea5-a41f-2b6614005b66">
<img width="1169" alt="Screenshot 2023-11-17 at 12 54 28" src="https://github.com/MetaMask/snaps/assets/68012282/15ab1051-745a-4bb8-ae69-331ab3f3ae99">

**Comment by Nick-1979** [comment]
Hello, this is Kami from Polkagate. We believe that adding the following components is essential:

1. A select component is crucial for the UI, especially for multi-chain platforms that require a chain-switching option for their users.
2. A skeleton component, which displays a loading indicator, would be highly beneficial during data retrieval or loading processes.

Thank you.
<img width="356" alt="Screenshot 2023-11-15 at 8 42 32 in the evening" src="https://github.com/MetaMask/snaps/assets/46442452/8dd26474-7079-4dca-8191-46f24ab69f06">

![ezgif com-video-to-gif](https://github.com/MetaMask/snaps/assets/46442452/9fe5a768-5f4d-4f60-9203-c5edc2f8874f)
